numpy/_core/_internal.py

```diff
@@ -0,0 +1,948 @@
+
+# Called in PyArray_DescrConverter function when
+#  a dictionary without "names" and "formats"
+#  fields is used as a data-type descriptor.
+def _usefields(adict, align):
+    try:
+        names = adict[-1]
+    except KeyError:
+        names = None
+    if names is None:
+        names, formats, offsets, titles = _makenames_list(adict, align)
+    else:
+        formats = []
+        offsets = []
+        titles = []
+        for name in names:
+            res = adict[name]
+            formats.append(res[0])
+            offsets.append(res[1])
+            if len(res) > 2:
+                titles.append(res[2])
+            else:
+                titles.append(None)
+
+    return dtype({"names": names,
+                  "formats": formats,
+                  "offsets": offsets,
+                  "titles": titles}, align)
+
+
+# construct an array_protocol descriptor list
+#  from the fields attribute of a descriptor
+# This calls itself recursively but should eventually hit
+#  a descriptor that has no fields and then return
+#  a simple typestring
+
+def _array_descr(descriptor):
+    fields = descriptor.fields
+    if fields is None:
+        subdtype = descriptor.subdtype
+        if subdtype is None:
+            if descriptor.metadata is None:
+                return descriptor.str
+            else:
+                new = descriptor.metadata.copy()
+                if new:
+                    return (descriptor.str, new)
+                else:
+                    return descriptor.str
+        else:
+            return (_array_descr(subdtype[0]), subdtype[1])
+
+    names = descriptor.names
+    ordered_fields = [fields[x] + (x,) for x in names]
+    result = []
+    offset = 0
+    for field in ordered_fields:
+        if field[1] > offset:
+            num = field[1] - offset
+            result.append(('', f'|V{num}'))
+            offset += num
+        elif field[1] < offset:
+            raise ValueError(
+                "dtype.descr is not defined for types with overlapping or "
+                "out-of-order fields")
+        if len(field) > 3:
+            name = (field[2], field[3])
+        else:
+            name = field[2]
+        if field[0].subdtype:
+            tup = (name, _array_descr(field[0].subdtype[0]),
+                   field[0].subdtype[1])
+        else:
+            tup = (name, _array_descr(field[0]))
+        offset += field[0].itemsize
+        result.append(tup)
+
+    if descriptor.itemsize > offset:
+        num = descriptor.itemsize - offset
+        result.append(('', f'|V{num}'))
+
+    return result
+
+# Build a new array from the information in a pickle.
+# Note that the name numpy._core._internal._reconstruct is embedded in
+# pickles of ndarrays made with NumPy before release 1.0
+# so don't remove the name here, or you'll
+# break backward compatibility.
+def _reconstruct(subtype, shape, dtype):
+    return ndarray.__new__(subtype, shape, dtype)
```

Comment:
This also should stay behind in `np.core`, I think. I don't know if there are any tests that exercise this but it would be nice to make sure this actually still works for really old pickles?

Reply:
Do you mean pickles that would use `numpy.core._internal._reconstruct`? 
I checked it and it refers to numpy versions before 1.0, as the comment says (for 1.0 it was already there for backward compatibility). So the last version where it's relevant is `0.9.8` from May 2006, whereas Python 3.0.0 was released in December 2008 (so such pickles would only be created with Python 2 that is EOL).

I tried to build numpy `0.9.8` with Python 2.7 conda environment but it complains:
```
AssertionError: hmm, why I am not inside SVN tree???
```

I looked through the codebase from that time and I didn't find any pickle files in test directories that I could take.

My solution to test the new location of legacy `_reconstruct`: 
I copied `numpy/lib/tests/data/py2-objarr.npy` to a new `py2-np0-objarr.npy` there and changed the path inside pickle to `numpy.core._internal._reconstruct` - this way the new location is tested. Would it be sufficient?

Reply:
One correction: Actually, there is a set of test for version 0 pickles! Here:
https://github.com/numpy/numpy/blob/67539a40cb13bad56a650809bf10a49e905a250d/numpy/core/tests/test_multiarray.py#L4329

So, the change of moving `_internal._reconstruct` is tested indeed.

I think that this additional file with a test can also stay.